app/scripts/ChromeSerialBridge.js

```diff
@@ -74,29 +75,49 @@ export default class ChromeSerialBridge {
       });
 
       serialPort.on('data', function (data) {
-        console.log('serialport data');
         var resp = {};
         resp.op = 'data';
         resp.data = data;
         port.postMessage(resp);
       });
 
-      port.onMessage.addListener(function (msg) {
-        console.log('socket received data');
+      function trySend(buffer) {
+        if (buffer) {
+          queue.push(buffer);
+        }
+
+        if (queue.length === 0) {
+          // Exhausted pending send buffer.
+          return;
+        }
 
-        // console.log('socket received', msg);
+        if (queue.length > 512) {
+          throw new Error('Send queue is full! More than 512 pending messages.');
+        }
+
+        var toSend = queue.shift();
+        serialPort.write(toSend, function (err, results) {
+          if (results.error) {
+            queue.unshift(toSend);
+          }
+
+          if (queue.length !== 0) {
+            setTimeout(trySend, 0);
```

Comment:
Is this the same as just calling trySend() recursively?

Reply:
Sort of! Since JS is single-threaded, any long running call can block the UI thread. Deferring with `setTimeout(..., 0)` or [`setImmediate` (nonstandard)](https://developer.mozilla.org/en-US/docs/Web/API/Window/setImmediate) tells the browser it can switch away from this and do something else, like update the UI or finish sending some serial data.

I'm not 100% sure this is necessary in an extension.

Reply:
My reason for adding it was to make sure we don't get caught in recursive `trySend()` calls while never pausing to allow the send to complete (in case it can get blocked by the JS thread).